Frontend/nix/watch.nix

```diff
@@ -0,0 +1,109 @@
+{ ... }:
+{
+  imports = [
+    ../node
+  ];
+
+  perSystem = { config, self', lib, pkgs, system, ... }:
+    let
+      webpack-dev-server = pkgs.writeShellApplication {
+        name = "webpack-dev-server-wrapped";
+        runtimeInputs = [ config.nammayatri.nodejs ];
+        text = ''
+          export NODE_PATH="${config.nammayatri.nodeDependencies}/node_modules"
+          PIPE="$1"
+          shift
+          tail -n1 -f "$PIPE" | node ${../node/watch.js} "$@"
+        '';
+      };
+
+      purifix-watch =
+        pkgs.writeShellApplication {
+          name = "purifix-watch";
+          runtimeInputs = [
+            pkgs.watchexec
+          ] ++ config.packages.ui-common.develop.buildInputs;
+          text = ''
+            PIPE=$1
+            shift
+            SRC_FOLDERS=()
+            SRC_GLOBS=()
+            for SRC_FOLDER in "$@"; do
+              if [ -d "$SRC_FOLDER/src" ]; then
+                SRC_FOLDERS+=( -w "$SRC_FOLDER/src" )
+                SRC_GLOBS+=( "\"$SRC_FOLDER/src/**/*.purs\"" )
+              fi
+            done
+            watchexec "''${SRC_FOLDERS[@]}" -r -f "**/*.purs" -- \
+              "echo 'suspend' >$PIPE;purifix ''${SRC_GLOBS[*]} && echo 'ready' >$PIPE"
+          '';
+        };
+
+      make-watch-processes = { target, env, platform }:
+        let
+          pipe = "/tmp/watch-${target}-${platform}-${env}";
+        in
+        {
+          processes = {
+            make-pipe = {
+              command = "mkfifo ${pipe}";
+              is_daemon = true;
+              shutdown.command = "rm -f ${pipe}";
+              availability.restart = "exit_on_failure";
+            };
+
+            purifix-watch = {
+              command = ''
+                cd ./Frontend/${target}
+                ${purifix-watch}/bin/purifix-watch ${pipe} . ../ui-common
+              '';
+              depends_on.make-pipe.condition = "process_started";
+            };
+
+            webpack-watch = {
+              command = ''
+                ${webpack-dev-server}/bin/webpack-dev-server-wrapped \
+                    ${pipe} \
+                    --env ${env} \
+                    --config $(realpath ./Frontend/${target}/webpack.${platform}.js) \
+                    --entry ./Frontend/${target}/index.js
+              '';
+              depends_on.make-pipe.condition = "process_started";
+            };
+          };
+        };
+
+      process-options =
+        lib.attrsets.cartesianProductOfSets {
+          env = [ "master" "sandbox" "prod" ];
+          target = [ "ui-customer" "ui-driver" ];
+          platform = [ "android" ]; # TODO: support iOS
+        };
```

Comment:
Looks like you want to factor out the "build config" separately, `./nix/build-config.nix` or something. The calculation of "name" happens in several places, and can be normalized here as well.